addons/binding/org.openhab.binding.dsmr/src/main/java/org/openhab/binding/dsmr/internal/DSMRHandlerFactory.java

```diff
@@ -0,0 +1,116 @@
+import org.openhab.binding.dsmr.handler.DSMRMeterHandler;
+import org.openhab.binding.dsmr.internal.discovery.DSMRMeterDiscoveryService;
+import org.openhab.binding.dsmr.internal.meter.DSMRMeterType;
+import org.osgi.framework.ServiceRegistration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link DSMRHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author M. Volaart
+ * @since 2.1.0
+ */
+public class DSMRHandlerFactory extends BaseThingHandlerFactory {
+    private final Logger logger = LoggerFactory.getLogger(DSMRHandlerFactory.class);
+
+    /**
+     * The registration handler used for notifying new detected meters
+     */
+    private ServiceRegistration<?> serviceReg;
+
+    /**
+     * Returns if the specified ThingTypeUID is supported by this handler.
+     *
+     * This handler support the THING_TYPE_DSMR_BRIDGE type and all ThingTypesUID that
+     * belongs to the supported DSMRMeterType objects
+     *
+     * @param {@link ThingTypeUID} to check
+     * @return true if the specified ThingTypeUID is supported, false otherwise
+     */
+    @Override
+    public boolean supportsThingType(ThingTypeUID thingTypeUID) {
+        if (thingTypeUID.equals(DSMRBindingConstants.THING_TYPE_DSMR_BRIDGE)) {
+            logger.debug("DSMR Bridge Thing {} supported", thingTypeUID);
+            return true;
+        } else {
+            boolean thingTypeUIDIsMeter = DSMRMeterType.METER_THING_TYPES.contains(thingTypeUID);
+            if (logger.isTraceEnabled()) {
+                if (thingTypeUIDIsMeter) {
+                    logger.trace("{} is a supported DSMR Meter thing", thingTypeUID);
+                } else {
+                    logger.trace("{} is not a DSMR Meter thing or not a supported DSMR Meter thing", thingTypeUID);
+                }
+            }
+            return thingTypeUIDIsMeter;
+        }
+    }
+
+    /**
+     * Create the ThingHandler for the corresponding Thing
+     *
+     * There are two handlers supported:
+     * - DSMRBridgeHandler that handle the Thing that corresponds to the physical DSMR device and does the serial
+     * communication
+     * - MeterHandler that handles the Meter things that are a logical part of the physical device
+     *
+     * @param thing The Thing to create a ThingHandler for
+     * @return ThingHandler for the given Thing or null if the Thing is not supported
+     */
+    @Override
+    protected ThingHandler createHandler(Thing thing) {
+        ThingTypeUID thingTypeUID = thing.getThingTypeUID();
+        logger.debug("Searching for thingTypeUID {}", thingTypeUID);
+        if (thingTypeUID.equals(DSMRBindingConstants.THING_TYPE_DSMR_BRIDGE)) {
+            Bridge dsmrBridge = (Bridge) thing;
+            DSMRMeterDiscoveryService discoveryService = new DSMRMeterDiscoveryService(dsmrBridge.getUID());
+            DSMRBridgeHandler bridgeHandler = new DSMRBridgeHandler((Bridge) thing, discoveryService);
```

Comment:
This is actually the wrong way around - a handler must not have a dependency on the discovery service - but a discovery service can use an existing handler. Check the HueThingHandlerFactory as an example: It passes the handler instance to the newly created discovery service:
```
HueLightDiscoveryService discoveryService = new HueLightDiscoveryService(bridgeHandler);
```